File: tools/_shared/utils/sanitize.py
"""
Input Sanitization Utilities for Python Applications
Provides security functions for validating and sanitizing user inputs
"""
import re
import os
from pathlib import Path
from typing import Any, Dict, List, Optional

def sanitize_filename(filename: str, max_length: int = 255) -> str:
    """
    Sanitize filename to prevent path traversal and invalid characters
    
    Args:
        filename: Original filename
        max_length: Maximum allowed length
        
    Returns:
        Sanitized filename safe for filesystem operations
    """
    # Remove any path components
    filename = os.path.basename(filename)
    
    # Remove or replace dangerous characters
    # Keep alphanumeric, dots, dashes, underscores
    filename = re.sub(r'[^\w\s.-]', '', filename)
    
    # Remove leading/trailing whitespace and dots
    filename = filename.strip('. ')
    
    # Replace multiple dots with single dot
    filename = re.sub(r'\.+', '.', filename)
    
    # Ensure filename is not empty
    if not filename:
        filename = 'unnamed_file'
    
    # Truncate to max length while preserving extension
    if len(filename) > max_length:
        name, ext = os.path.splitext(filename)
        max_name_length = max_length - len(ext)
        filename = name[:max_name_length] + ext
    
    return filename


def validate_file_type(filename: str, allowed_extensions: List[str]) -> bool:
    """
    Validate file extension against allowed types
    
    Args:
        filename: File name to validate
        allowed_extensions: List of allowed extensions (e.g., ['xlsx', 'csv'])
        
    Returns:
        True if file type is allowed, False otherwise
    """
    if not filename or '.' not in filename:
        return False
    
    ext = filename.rsplit('.', 1)[1].lower()
    return ext in [e.lower() for e in allowed_extensions]


def validate_file_size(file_size: int, max_size_mb: int = 100) -> bool:
    """
    Validate file size against maximum allowed size
    
    Args:
        file_size: File size in bytes
        max_size_mb: Maximum allowed size in megabytes
        
    Returns:
        True if size is within limit, False otherwise
    """
    max_size_bytes = max_size_mb * 1024 * 1024
    return file_size <= max_size_bytes


def sanitize_path(path: str, base_dir: str) -> Optional[str]:
    """
    Sanitize and validate path to prevent directory traversal
    
    Args:
        path: Path to sanitize
        base_dir: Base directory that path must be within
        
    Returns:
        Sanitized absolute path if valid, None if path traversal detected
    """
    # Resolve to absolute path
    abs_path = os.path.abspath(os.path.join(base_dir, path))
    abs_base = os.path.abspath(base_dir)
    
    # Check if path is within base directory
    if not abs_path.startswith(abs_base):
        return None
    
    return abs_path


def escape_html(text: str) -> str:
    """
    Escape HTML special characters to prevent XSS
    
    Args:
        text: Text to escape
        
    Returns:
        HTML-escaped text
    """
    html_escape_table = {
        "&": "&amp;",
        '"': "&quot;",
        "'": "&#x27;",
        ">": "&gt;",
        "<": "&lt;",
    }
    return "".join(html_escape_table.get(c, c) for c in text)


def sanitize_dict(data: Dict[str, Any], remove_null: bool = True) -> Dict[str, Any]:
    """
    Sanitize dictionary by removing null values and dangerous patterns
    
    Args:
        data: Dictionary to sanitize
        remove_null: Whether to remove None/null values
        
    Returns:
        Sanitized dictionary
    """
    sanitized = {}
    
    for key, value in data.items():
        # Skip None values if configured
        if remove_null and value is None:
            continue
        
        # Recursively sanitize nested dicts
        if isinstance(value, dict):
            sanitized[key] = sanitize_dict(value, remove_null)
        # Recursively sanitize lists
        elif isinstance(value, list):
            sanitized[key] = [
                sanitize_dict(item, remove_null) if isinstance(item, dict) else item
                for item in value
            ]
        else:
            sanitized[key] = value
    
    return sanitized


def validate_upload_file(file, allowed_extensions: List[str], max_size_mb: int = 100) -> tuple:
    """
    Comprehensive file upload validation
    
    Args:
        file: File object from request
        allowed_extensions: List of allowed file extensions
        max_size_mb: Maximum file size in MB
        
    Returns:
        Tuple of (is_valid, error_message, sanitized_filename)
    """
    # Check if file exists
    if not file or not file.filename:
        return False, "No file provided", None
    
    # Sanitize filename
    safe_filename = sanitize_filename(file.filename)
    
    # Validate file extension
    if not validate_file_type(safe_filename, allowed_extensions):
        return False, f"Invalid file type. Allowed: {', '.join(allowed_extensions)}", None
    
    # Check file size if available
    if hasattr(file, 'content_length') and file.content_length:
        if not validate_file_size(file.content_length, max_size_mb):
            return False, f"File too large. Maximum size: {max_size_mb}MB", None
    
    return True, None, safe_filename


def normalize_sku(sku: str) -> str:
    """
    Normalize SKU format for consistency
    
    Args:
        sku: Product SKU
        
    Returns:
        Normalized SKU (uppercase, trimmed)
    """
    if not sku:
        return ""
    
    # Convert to uppercase and remove whitespace
    normalized = str(sku).upper().strip()
    
    # Remove special characters except dashes and underscores
    normalized = re.sub(r'[^\w-]', '', normalized)
    
    return normalized

File: tools/_shared/utils/sanitize.js
/**
 * Input Sanitization Utilities
 * Prevents SQL injection, XSS, and other injection attacks
 */

/**
 * Sanitize string for use in MongoDB regex
 * Prevents NoSQL injection attacks
 * @param {string} input - User input to sanitize
 * @returns {string} Sanitized string safe for regex
 */
function sanitizeForRegex(input) {
  if (!input || typeof input !== 'string') {
    return '';
  }
  // Escape special regex characters
  return input.replace(/[.*+?^${}()|[\]\\]/g, '\\$&').trim();
}

/**
 * Sanitize SKU input
 * @param {string} sku - SKU to sanitize
 * @returns {string} Sanitized SKU
 */
function sanitizeSku(sku) {
  if (!sku || typeof sku !== 'string') {
    return '';
  }
  // Remove special characters, keep alphanumeric, hyphen, underscore
  return sku.replace(/[^a-zA-Z0-9\-_]/g, '').trim().toUpperCase();
}

/**
 * Sanitize email address
 * @param {string} email - Email to sanitize
 * @returns {string|null} Sanitized email or null if invalid
 */
function sanitizeEmail(email) {
  if (!email || typeof email !== 'string') {
    return null;
  }
  
  const trimmed = email.trim().toLowerCase();
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  
  return emailRegex.test(trimmed) ? trimmed : null;
}

/**
 * Sanitize phone number
 * @param {string|number} phone - Phone number to sanitize
 * @returns {string|null} Normalized phone number or null
 */
function sanitizePhone(phone) {
  if (!phone) {
    return null;
  }
  
  const phoneStr = String(phone).trim();
  if (!phoneStr || phoneStr.toLowerCase() === 'nan' || phoneStr === 'null') {
    return null;
  }
  
  // Extract only digits
  const digits = phoneStr.replace(/\D/g, '');
  
  // Must have at least 7 digits
  if (digits.length < 7) {
    return null;
  }
  
  // For international numbers, keep last 10 digits
  if (digits.length > 10) {
    return digits.slice(-10);
  }
  
  return digits;
}

/**
 * Sanitize filename for safe file system operations
 * @param {string} filename - Original filename
 * @returns {string} Safe filename
 */
function sanitizeFilename(filename) {
  if (!filename || typeof filename !== 'string') {
    return 'unnamed_file';
  }
  
  // Remove path traversal attempts
  let safe = filename.replace(/\.\./g, '');
  
  // Remove or replace unsafe characters
  safe = safe.replace(/[<>:"|?*\x00-\x1f]/g, '_');
  
  // Remove leading/trailing dots and spaces
  safe = safe.replace(/^[\s.]+|[\s.]+$/g, '');
  
  // Limit length
  if (safe.length > 255) {
    const ext = safe.split('.').pop();
    const name = safe.slice(0, 255 - ext.length - 1);
    safe = `${name}.${ext}`;
  }
  
  return safe || 'unnamed_file';
}

/**
 * Sanitize object for MongoDB query
 * Removes properties starting with $ to prevent operator injection
 * @param {Object} obj - Object to sanitize
 * @returns {Object} Sanitized object
 */
function sanitizeMongoQuery(obj) {
  if (!obj || typeof obj !== 'object') {
    return obj;
  }
  
  if (Array.isArray(obj)) {
    return obj.map(sanitizeMongoQuery);
  }
  
  const sanitized = {};
  for (const [key, value] of Object.entries(obj)) {
    // Skip keys starting with $ (MongoDB operators)
    if (key.startsWith('$')) {
      continue;
    }
    
    // Recursively sanitize nested objects
    if (value && typeof value === 'object') {
      sanitized[key] = sanitizeMongoQuery(value);
    } else {
      sanitized[key] = value;
    }
  }
  
  return sanitized;
}

/**
 * Backwards-compatible sanitizer used by shared middleware.
 * Currently mirrors sanitizeMongoQuery but can be extended for other rules.
 * @param {Object} obj
 * @returns {Object}
 */
function sanitizeObject(obj) {
  return sanitizeMongoQuery(obj);
}

/**
 * Validate and sanitize URL
 * @param {string} url - URL to validate
 * @returns {string|null} Valid URL or null
 */
function sanitizeUrl(url) {
  if (!url || typeof url !== 'string') {
    return null;
  }
  
  try {
    const parsed = new URL(url);
    // Only allow http and https
    if (!['http:', 'https:'].includes(parsed.protocol)) {
      return null;
    }
    return parsed.toString();
  } catch {
    return null;
  }
}

/**
 * Sanitize HTML to prevent XSS
 * Basic implementation - use a library like DOMPurify for production
 * @param {string} html - HTML string to sanitize
 * @returns {string} Sanitized HTML
 */
function sanitizeHtml(html) {
  if (!html || typeof html !== 'string') {
    return '';
  }
  
  return html
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;')
    .replace(/\//g, '&#x2F;');
}

module.exports = {
  sanitizeForRegex,
  sanitizeSku,
  sanitizeEmail,
  sanitizePhone,
  sanitizeFilename,
  sanitizeMongoQuery,
  sanitizeObject,
  sanitizeUrl,
  sanitizeHtml
};
